src/Stratis.Bitcoin/Base/ConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,434 @@
+                return;
+            }
+
+            var headers = new HeadersPayload();
+            ChainedHeader consensusTip = this.consensusManager.Tip;
+
+            ChainedHeader fork = this.chain.FindFork(getHeadersPayload.BlockLocators);
+            if (fork != null)
+            {
+                if ((consensusTip == null) || (fork.Height > consensusTip.Height))
+                {
+                    // Fork not yet validated.
+                    fork = null;
+                }
+
+                if (fork != null)
+                {
+                    foreach (ChainedHeader header in this.chain.EnumerateToTip(fork).Skip(1))
+                    {
+                        if (header.Height > consensusTip.Height)
+                            break;
+
+                        headers.Headers.Add(header.Header);
+                        if ((header.HashBlock == getHeadersPayload.HashStop) || (headers.Headers.Count == 2000))
+                            break;
+                    }
+                }
+            }
+
+            await peer.SendMessageAsync(headers).ConfigureAwait(false);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Processes "headers" message received from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="headersPayload">Payload of "headers" message to process.</param>
+        /// <remarks>
+        /// "headers" message is sent in response to "getheaders" message or it is solicited
+        /// by the peer when a new block is validated (unless in IBD).
+        /// <para>
+        /// When we receive "headers" message from the peer, we can adjust our knowledge
+        /// of the peer's view of the chain. We update its pending tip, which represents
+        /// the tip of the best chain we think the peer has.
+        /// </para>
+        /// <para>
+        /// If we receive a valid header from peer which work is higher than the work
+        /// of our best chain's tip, we update our view of the best chain to that tip.
+        /// </para>
+        /// </remarks>
+        private async Task ProcessHeadersAsync(INetworkPeer peer, HeadersPayload headersPayload)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(peer), peer.RemoteSocketEndpoint, nameof(headersPayload), headersPayload);
+
+            List<BlockHeader> headers = headersPayload.Headers;
+
+            if (headers.Count == 0)
+            {
+                this.logger.LogTrace("Headers payload with no headers was received. Assuming we're synced with the peer.");
+                this.logger.LogTrace("(-)[NO_HEADERS]");
+                return;
+            }
+
+            // Check headers for consecutiveness.
+            for (int i = 1; i < headers.Count; ++i)
+            {
+                if (headers[i].HashPrevBlock != headers[i - 1].GetHash())
```

Comment:
Cant you rewrite this into a `foreach` and include the condition into the `where(x=>x...)`?

Reply:
Here we have only BlockHeader instances so there is no way to access prev header inside `where(x=> x...)`